### PR TITLE
SRC-12 Contract Factory Standard

### LINKED ISSUE #52: SRC-11; Contract Factory Standard
### Motivation

# Abstract

The following standard allows for the implementation of a standard ABI for Contract Factories using the Sway Language. The standardized design designates how verification of newly deployed child contracts are handled.

# Motivation

A standard interface for Contract Factories provides a safe and effective method of ensuring contracts can verify the validity of another contract as a child of a factory. This is critical on the Fuel Network as contracts cannot deploy other contracts and verification must be done after deployment.

# Prior Art

A Contract Factory is a design where a template contract is used and deployed repeatedly with different configurations. These configurations are often minor changes such as pointing to a different asset. All base functionality remains the same.

On Fuel, contracts cannot deploy other contracts. As a result, a Contract Factory on Fuel must register and verify that the bytecode root of a newly deployed child contract matches the expected bytecode root.

When changing something such as a configurable in Sway, the bytecode root is recalculated. As such any differences between child contracts must be made in storage and a constructor/initialization function is required. [This](https://github.com/FuelLabs/fuel-specs/issues/490) issue investigates whether an opcode would allow for the verification of a child contract with differentiating configurable values.

# Specification

The following functions MUST be implemented to follow the SRC-11; Contract Factory Standard:

### - `fn register_contract(child_contract: ContractId)`

The `register_contract()` function verifies that a newly deployed contract is the child of a contract factory.

- This function MUST verify that the bytecode root of the `child_contract` contract matches the expected bytecode root.
- This function MAY add arbitrary conditions checking a contract factory child’s validity, such as verifying storage variables or initialized values.

### - `fn is_valid(child_contract: ContractId) -> bool`

The `is_valid()` function returns a boolean representing the state of whether a contract is a valid child of the contract factory. 

- This function MUST return `true` if this is a valid child, otherwise `false`.

### - `fn factory_bytecode_root() -> Option<b256>`

The `factory_bytecode_root()` function returns the bytecode root of the default template contract.

- This function MUST return the bytecode root used to verify contracts against.

## Optional Functions

The following are functions that may enhance the use of the SRC-11 standard but ARE NOT required. Function names MUST be followed, however function arguments MAY differentiate between contracts.

### - `fn get_contract(arguments: Bytes) -> Option<ContractId>`

The `get_contract()` function will return a registered contract factory child contract with specific implementation details specified by `arguments`.

- This function MUST return `Some(ContractId)` IF a contract that follows the specified `arguments` has been registered with the SRC-11 Contract Factory contract, otherwise `None`.

# Rationale 

The SRC-11; Contract Factory Standard is designed to standardize the contract factory design implementation interface between all Fuel instances. 

# Backwards Compatibility

There are no other standards that the SRC-11 requires compatibility.

# Example ABI

```sway
abi SRC11 {
     #[storage(read, write)]
     fn register_contract(child_contract: ContractId);
     fn is_valid(child_contract: ContractId) -> bool;
     #[storage(read)]
     fn factory_bytecode_root() -> Option<b256>;
}

abi SRC11_Extension {
     fn get_contract(arguments: Bytes) -> ContractId;
}
```

### PR DESCRIPTION
**Pull request by bitzoic**
## Type of change

- New Standard

## Changes

The following changes have been made:

- Adds the SRC-12 Contract Factory Standard Specification
- Adds the SRC-12 Contract Factory Standard ABI
- Adds the SRC-12 Contract Factory Standard Examples

## Related Issues

Closes #52

### CODE REVIEW

File: standards/src12-contract-factory/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+The following functions MUST be implemented to follow the SRC-12; Contract Factory Standard:
+
+## Required Functions
+
+### - `fn register_contract(child_contract: ContractId, configurables: Option<Vec<(u64, Vec<u8>)>>)`
+
+The `register_contract()` function verifies that a newly deployed contract is the child of a contract factory.
+
+- This function MUST verify that the bytecode root of the `child_contract` contract matches the expected bytecode root.
```
Comment (SwayStar123) [inline standards/src12-contract-factory/README.md:34]: should this not specify what the expected bytecode root is?

File: standards/src12-contract-factory/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+When changing something such as a configurable in Sway, the bytecode root is recalculated. The [Bytecode Library](https://github.com/FuelLabs/sway-libs/tree/master/libs/bytecode) has been developed to calculate the bytecode root of a contract with different configurables.
+
+# Specification
+
+The following functions MUST be implemented to follow the SRC-12; Contract Factory Standard:
+
+## Required Functions
+
+### - `fn register_contract(child_contract: ContractId, configurables: Option<Vec<(u64, Vec<u8>)>>)`
```
Comment (SwayStar123) [inline standards/src12-contract-factory/README.md:30]: I dont see any statements saying it should revert it failed, assuming thats intentional I think this should return a bool, to indicate whether the registration succeeded or failed, essentially save a is_valid call.

File: standards/src12-contract-factory/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+- This function MUST return `true` if this is a valid and registered child, otherwise `false`.
+
+### - `fn factory_bytecode_root() -> Option<b256>`
+
+The `factory_bytecode_root()` function returns the bytecode root of the default template contract.
+
+- This function MUST return the bytecode root of the template contract.
+
+### - `fn get_contract_id(configurables: Option<Vec<(u64, Vec<u8>)>>) -> Option<ContractId>`
```
Comment (SwayStar123) [inline standards/src12-contract-factory/README.md:50]: this function name is not very indicative of its function

Comment (SwayStar123) [reply standards/src12-contract-factory/README.md:50]: Also below this function seems to be part of the "Extension" abi, so this is not really a required function is it?

File: standards/src12-contract-factory/README.md (hunk 1 above)
Comment (bitzoic) [reply standards/src12-contract-factory/README.md:34]: Not sure I follow. Expected is whatever you anticipate it is. That can be something you calculate or pre-determined

File: standards/src12-contract-factory/README.md (hunk 3 above)
Comment (bitzoic) [reply standards/src12-contract-factory/README.md:50]: How is `get_contract_id` not indicative of what it's intended to do?? 

This is an extension as it's impossible to get the contract id if there are no configurables. Many contracts all with the same bytecode root and no way to differentiate them

**Review by SwayStar123 — APPROVED** [review]

**Review by SwayStar123 — APPROVED** [review]

**Review by SwayStar123 — DISMISSED** [review]

File: examples/src12-contract-factory/with_configurables/src/with_configurables.sw [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            .insert(sha256(configurables.unwrap_or(Vec::new())), child_contract);
+
+        return Result::Ok(computed_root)
+    }
+
+    /// Returns a boolean representing the state of whether a contract is a valid child of the contract factory.
+    ///
+    /// # Arguments
+    ///
+    /// * `child_contract`: [ContractId] - The deployed factory child contract which to check the registry status.
```
Comment (K1-R1) [inline examples/src12-contract-factory/with_configurables/src/with_configurables.sw:116]:
```suggestion
    /// * `child_contract`: [ContractId] - The deployed factory child contract of which to check the registry status.
```

File: examples/src12-contract-factory/without_configurables/src/without_configurables.sw [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+impl SRC12 for Contract {
+    /// Verifies that a newly deployed contract is the child of a contract factory and registers it.
+    ///
+    /// # Additional Information
+    ///
+    /// This example does not check whether a contract has already been registered and will overwrite any values.
+    ///
+    /// # Arguments
+    ///
+    /// * `child_contract`: [ContractId] - The deployed factory child contract which to verify the bytecode root.
```
Comment (K1-R1) [inline examples/src12-contract-factory/without_configurables/src/without_configurables.sw:24]:
```suggestion
    /// * `child_contract`: [ContractId] - The deployed factory child contract of which to verify the bytecode root.
```

File: examples/src12-contract-factory/without_configurables/src/without_configurables.sw [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+        storage.registered_contracts.insert(child_contract, true);
+        return Result::Ok(returned_root)
+    }
+
+    /// Returns a boolean representing the state of whether a contract is a valid child of the contract factory.
+    ///
+    /// # Arguments
+    ///
+    /// * `child_contract`: [ContractId] - The deployed factory child contract which to check the registry status.
```
Comment (K1-R1) [inline examples/src12-contract-factory/without_configurables/src/without_configurables.sw:72]:
```suggestion
    /// * `child_contract`: [ContractId] - The deployed factory child contract of which to check the registry status.
```

File: standards/src12-contract-factory/src/src12.sw [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+pub type BytecodeRoot = b256;
+pub type ContractConfigurables = Vec<(u64, Vec<u8>)>;
+
+abi SRC12 {
+    /// Verifies that a newly deployed contract is the child of a contract factory and registers it.
+    ///
+    /// # Arguments
+    ///
+    /// * `child_contract`: [ContractId] - The deployed factory child contract which to verify the bytecode root.
```
Comment (K1-R1) [inline standards/src12-contract-factory/src/src12.sw:13]:
```suggestion
    /// * `child_contract`: [ContractId] - The deployed factory child contract of which to verify the bytecode root.
```

File: standards/src12-contract-factory/src/src12.sw [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    fn register_contract(
+        child_contract: ContractId,
+        configurables: Option<ContractConfigurables>,
+    ) -> Result<BytecodeRoot, str>;
+
+    /// Returns a boolean representing the state of whether a contract is a valid child of the contract factory.
+    ///
+    /// # Arguments
+    ///
+    /// * `child_contract`: [ContractId] - The deployed factory child contract which to check the registry status.
```
Comment (K1-R1) [inline standards/src12-contract-factory/src/src12.sw:41]:
```suggestion
    /// * `child_contract`: [ContractId] - The deployed factory child contract of which to check the registry status.
```

**Review by K1-R1 — COMMENTED** [review]
Looks good. Just minor comments added. Couldn't decide on `of which` or `for which` in my suggestion.

**Review by K1-R1 — DISMISSED** [review]

**Review by K1-R1 — APPROVED** [review]

**Review by SwayStar123 — APPROVED** [review]

**Pull request merged**